### ISSUE TITLE
sql: logictest leaks memory

### ISSUE THREAD

**Issue by cucaroach** [cockroachdb/cockroach#78690]
```
make test 'TESTFLAGS=-memprofile=mem.out vv -bigtest -flex-types' TESTTIMEOUT=24h PKG=./pkg/sql/logictest 'TESTS=^TestSqlLiteLogic/local/slt_good_0.test' TAGS=crdb_test_off
```

```
❯ go tool pprof mem.out
File: logictest.test
Build ID: f25562c981de7971c6ec0341c073a16fcd6f0ac0
Type: alloc_space
Time: Mar 28, 2022 at 5:21pm (UTC)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 522.93GB, 45.48% of 1149.78GB total
Dropped 6760 nodes (cum <= 5.75GB)
Showing top 10 nodes out of 485
      flat  flat%   sum%        cum   cum%
  185.38GB 16.12% 16.12%   221.53GB 19.27%  github.com/cockroachdb/cockroach/pkg/util/pretty.(*beExec).iDoc
  116.63GB 10.14% 26.27%   357.03GB 31.05%  github.com/cockroachdb/cockroach/pkg/util/pretty.(*beExec).be
   52.52GB  4.57% 30.84%    52.52GB  4.57%  github.com/cockroachdb/pebble/vfs.(*memFile).Write
   36.15GB  3.14% 33.98%    36.15GB  3.14%  github.com/cockroachdb/cockroach/pkg/util/pretty.(*beExec).newiDoc (inline)
   28.64GB  2.49% 36.47%    28.64GB  2.49%  github.com/cockroachdb/cockroach/pkg/storage.(*pebbleResults).put
   23.43GB  2.04% 38.51%    23.43GB  2.04%  reflect.unsafe_NewArray
   21.25GB  1.85% 40.36%    21.25GB  1.85%  github.com/cockroachdb/cockroach/pkg/util/pretty.(*beExec).newDocBest (inline)
   20.15GB  1.75% 42.11%    25.46GB  2.21%  github.com/cockroachdb/cockroach/pkg/sql/parser.(*Parser).scanOneStmt
   19.44GB  1.69% 43.80%    19.44GB  1.69%  reflect.New
   19.33GB  1.68% 45.48%    19.33GB  1.68%  github.com/cockroachdb/cockroach/pkg/sql/opt/constraint.SingleSpanConstraint
```
![profile001](https://user-images.githubusercontent.com/80277990/160512480-989e5efd-0172-4f63-81a0-bb0f9d1a9127.svg)

Jira issue: CRDB-14218

**Comment by ajwerner** [comment]
Can you say more about the leak? This profile is showing `alloc_space`, which is all space allocated over the profile's time period. Have a look at `inuse_space` for the stuff that is currently retained.

**Comment by cucaroach** [comment]
So I'm assuming this is a memory leak because it eventually causes my machine to run out of memory and fall over.  I narrowed down the scope of the test to get it to finish and get a memprofile.   But I think the way I ran logictest maybe fails to catch the problem because heap profile is presumably a snapshot of where we are at when the process exits?     Do I need to add a periodic heap sampling feature to logictest to find out what the inuse objects are that are driving heap expansions?

**Comment by ajwerner** [comment]
> Do I need to add a periodic heap sampling feature to logictest to find out what the inuse objects are that are driving heap expansions?

Might be worthwhile. https://github.com/cockroachdb/cockroach/blob/a0a02fb9e8a8e2613529dc61ef50c59a8a0dbd5b/pkg/server/heapprofiler/heapprofiler.go#L44 + a goroutine polling the thing might be helpful.

**Comment by cucaroach** [comment]
On CI it appears to succeed with 13GB of memory usage:

```
I220323 11:30:46.975626 246610898 2@server/status/runtime_log.go:47 â‹® [n2] 146363  runtime stats: 13 GiB RSS, 969 goroutines (stacks: 18 MiB), 920 MiB/8.4 GiB Go alloc/total(stale) (heap fragmentation: 1.1 GiB, heap reserved: 4.8 GiB, heap released: 6.1 GiB), 72 MiB/98 MiB CGO alloc/total (0.3 CGO/sec), 192.5/12.4 %(u/s)time, 0.0 %gc (13706x), 60 MiB/60 MiB (r/w)net
```
